readme.md
REDIPS.dialog 1.8.0
============

## What id REDIPS.dialog?

REDIPS.dialog is simple JavaScript dialog box created with two DIV elements. First DIV overlays whole page
(transparency is styled to 60%) while second DIV (dialog box) is positioned at page center.

## Features

* dialog with one or two buttons
* actions on button click (user functions calls)
* display images in dialog box
* combining images and text
* show dialog from HTML page (HTML, PHP, ASPX ...)
* show custom HTML
* display YouTube video in dialog box
* event.displayed() and event.closed() event handlers invoked after dialog is displayed / closed

## Public methods

* REDIPS.dialog.init() - dialog box initialization
* REDIPS.dialog.show() - show dialog box
* REDIPS.dialog.hide() - hide dialog box
* REDIPS.dialog.html() - define custom HTML to display in dialog box

## Demos

Demo page contains several examples of REDIPS.dialog box. 

* [http://www.redips.net/javascript/dialog-box/](http://www.redips.net/javascript/dialog-box/)


changes.txt
Changes for REDIPS.dialog

1.8.0 (2020-03-24)
- REDIPS.drag is checked with ESLint - cleaning and optimization (deleted unused variables)
- myhandler_displayed() changed to event.displayed()
- myhandler_closed() changed to event.closed()
- internal AJAX call updated
- displaying YouTube in new format "youtube:{youtubeId}" - REDIPS.dialog.show(640, 390, 'youtube:2JWaWfKc5Vg')
- changed YouTube embeded HTML code

1.7.0
- added REDIPS.dialog.html() method (defines custom HTML to show in dialog box)
- added example01 to show how to display custom HTML in dialog box
- redips-dialog-min.js is compressed with using Google Closure Compiler
- fixed bug when page is not found
- comments in redips-dialog.js are written in JSDoc style

1.6.0
- added ASPX page definition to fetch via AJAX -> REDIPS.dialog.show(520, 400, 'Email.aspx')
- added event handlers myhandler_displayed() and myhandler_closed()
- added example00 to show how to use new event handlers

1.5.6
- added trim for function name and function parameter
  * space characters will not cause error ('Yes| button2 ' will work as well as 'Yes|button2')

1.5.5
- added REDIPS.dialog.shown public property

1.5.4
- fixed missing quote - thanks to Bianka Martinovic

1.5.3
- HTML tags converted to lowercase (XHTML requires HTML tags to be lowercase) - thanks to Bianka Martinovic

1.5.2
- added AJAX request for HTML/PHP page (instead of iframe)

1.5.1
- added public property REDIPS.dialog.close_button to change default '✕' character if needed

1.5.0
- added support to show dialog from html page

1.4.4
 - improved REDIPS.event (fixed bug with adding events in IE8 - appeared in REDIPS.drag)

1.4.3
- switched to player version 2
- changed skin color
- redesigned close button

1.4.2
- closing input tag in line 172 due to breaking xhtml (thanks to cs0815)
- tag was forced to lowercase because styles were never applied (thanks to cs0815)

1.4.1
- REDIPS namespace created if is not already defined in another REDIPS package (var REDIPS = REDIPS || {};)
- added REDIPS.event at the bottom

1.4.0
- <div id="dialog"> changed to <div id="REDIPSdialog">
- button "Close" moved to title bar
- added support for YouTube
- dialog box position changed to "fixed" (instead of "absolute")
- class names renamed to avoid collision

1.3.0
- fixed bug with properties op_high, op_low and fade_speed
- enabled passing parameter to the user function


header.js
/* eslint-env browser */
/* eslint
   semi: ["error", "always"],
   indent: [2, "tab"],
   no-tabs: 0,
   no-multiple-empty-lines: ["error", {"max": 2, "maxEOF": 1}],
   one-var: ["error", "always"],
   no-trailing-spaces: ["error", { "ignoreComments": true }] */

/* enable strict mode */
'use strict';

// global variables
var redipsURL = redipsURL || '/javascript/dialog-box/', // eslint-disable-line no-use-before-define
	headerInit;

// header initialization
headerInit = function () {
	let header = document.createElement('div'),
		title = document.title,
		href = window.location.href.split('/'),
		indexLink = '<a href="..">index</a>';
	// index link is not needed for main page
	if (href[href.length - 2].startsWith('REDIPS')) {
		indexLink = '';
	}
	// add "header" DIV element
	document.body.insertBefore(header, document.body.firstChild);
	// apply inner HTML
	header.innerHTML = '<div style="color:white;background-color:#1e73be;padding:10px;text-align:center;font-size:20px">' + title + '</div>' +
		'<div style="float:left;width:50%;padding-left:10px"><a href="http://www.redips.net' + redipsURL + '">www.redips.net</a></div>' +
		'<div style="text-align:right;padding-right:10px;margin-bottom:10px">' + indexLink + '</div>';
};

// add onload event listener
if (window.addEventListener) {
	window.addEventListener('load', headerInit, false);
}
else if (window.attachEvent) {
	window.attachEvent('onload', headerInit);
}


script.js
/* eslint-env browser */
/* eslint
   semi: ["error", "always"],
   indent: [2, "tab"],
   no-tabs: 0,
   no-multiple-empty-lines: ["error", {"max": 2, "maxEOF": 1}],
   one-var: ["error", "always"] */
/* global REDIPS */

/* enable strict mode */
'use strict';

// create redips container
var redips = {};


// dialog box initialization (called from onload event)
redips.init = function () {
	REDIPS.dialog.init();
	REDIPS.dialog.op_high = 60;
	REDIPS.dialog.fade_speed = 18;
	// REDIPS.dialog.close_button = 'Z';
	// if 'html' string is send as parameter in REDIPS.dialog.show() then this HTML will be used
	REDIPS.dialog.html('<div><a href="http://www.redips.net/" title="REDIPS site">www.redips.net</a><br/><br/>This is my HTML</div>');
};


// execute after button1 is clicked
function button1 () { // eslint-disable-line no-unused-vars
	alert('Hi from function button1!');
};


// execute after button2 is clicked
function button2 (param) { // eslint-disable-line no-unused-vars
	var msg = 'Hi from function button2!';
	// if parameter exists
	if (param !== undefined) {
		msg += '\nParameter: ' + param;
	}
	alert(msg);
};


// add onload event listener
if (window.addEventListener) {
	window.addEventListener('load', redips.init, false);
}
else if (window.attachEvent) {
	window.attachEvent('onload', redips.init);
}


redips-dialog-min.js
/*
Copyright (c)  2008-2020, www.redips.net  All rights reserved.
Code licensed under the BSD License: http://www.redips.net/license/
http://www.redips.net/javascript/dialog-box/
Version 1.8.0
Mar 24, 2020.
*/
var REDIPS=REDIPS||{};
REDIPS.dialog=function(){var c,g,e,f,p=0,n=0,l,w;var m=function(a,b,d){void 0===b&&(b="");void 0===d&&(d="");f.innerHTML='<div class="redips-dialog-titlebar"><span title="Close" onclick="REDIPS.dialog.hide(\'undefined\')">'+REDIPS.dialog.closeButton+'</span></div><table class="redips-dialog-tbl" cellpadding="0" cellspacing="0"><tr><td valign="center" height="'+n+'" width="'+p+'">'+a+'<div class="redips-dialog-buttons">'+b+d+"</div></td></tr></table>";e.style.display=f.style.display="block";t(REDIPS.dialog.opLow,
5)};var x=function(a){a=a.split("|");var b=a[2];b=void 0!==b?"','"+b:"";return'<input type="button" onclick="REDIPS.dialog.hide(\''+u(a[1])+u(b)+'\');" value="'+u(a[0])+'"/>'};var y=function(a){var b;var d=a.split(",");if(1===d.length)a='<div class="redips-dialog-imgc"><img src="'+d[0]+'" height="'+(n-40)+'"/></div>';else{a='<div class="redips-dialog-imgc" style="width:'+(p-8)+'px"><table><tr>';for(b=0;b<d.length;b++)a+='<td><img src="'+d[b]+'" height="'+(n-40)+'"/></td>';a+="</tr></table></div>"}return a};
var v=function(){if("number"===typeof window.innerWidth){var a=window.innerWidth;var b=window.innerHeight;var d=window.pageXOffset;var c=window.pageYOffset}else document.documentElement&&(document.documentElement.clientWidth||document.documentElement.clientHeight)?(a=document.documentElement.clientWidth,b=document.documentElement.clientHeight,d=document.documentElement.scrollLeft,c=document.documentElement.scrollTop,e.style.width=a+"px",e.style.height=b+"px"):document.body&&(document.body.clientWidth||
document.body.clientHeight)&&(a=document.body.clientWidth,b=document.body.clientHeight,d=document.body.scrollLeft,c=document.body.scrollTop);f.style.left=(a-p)/2+"px";f.style.top=(b-n)/2+"px";e.style.top=c+"px";e.style.left=d+"px"};var t=function(a,b){e.style.opacity=a/100;e.style.filter="alpha(opacity="+a+")";a+=b;if(REDIPS.dialog.opLow<=a&&a<=REDIPS.dialog.opHigh)setTimeout(function(){t(a,b)},REDIPS.dialog.fadeSpeed);else if(REDIPS.dialog.opLow>a&&(e.style.display="none","undefined"!==l))window[l](w)};
var u=function(a){void 0!==a&&(a=a.replace(/^\s+|\s+$/g,"").replace(/\s{2,}/g," "));return a};return{opHigh:60,opLow:0,fadeSpeed:10,youtube:'<iframe width="640" height="390" src="https://www.youtube.com/embed/_youtubeId_" frameborder="0" allowfullscreen></iframe>',shown:!1,closeButton:"\u2715",init:function(){f=document.createElement("div");f.setAttribute("id","redips-dialog");g=document.createElement("div");e=document.createElement("div");e.setAttribute("id","redips-dialog-shade");var a=document.getElementsByTagName("body")[0];
a.appendChild(e);a.appendChild(f);REDIPS.event.add(window,"resize",v);REDIPS.event.add(window,"scroll",v);c=new XMLHttpRequest},show:function(a,b,d,e,f){var q="",r="",h="",l=h="";var k="";REDIPS.dialog.shown=!0;p=a;n=b;v();a=/(\.jpg|\.jpeg|\.gif|\.png)$/i;b=/(\.php|\.html|\.aspx)/i;k=/youtube:/i;void 0!==e&&(q=x(e));void 0!==f&&(r=x(f));"html"===d?m(g.innerHTML,q,r):a.test(d)?(k=d.split("|"),1===k.length?h=y(k[0]):(h=y(k[1]),l="<div>"+k[0]+"</div>"),m(h+l,q,r)):b.test(d)?(c.open("GET",d,!0),c.onreadystatechange=
function(){c.readyState===XMLHttpRequest.DONE&&(200===c.status?m(c.responseText):m("Error: ["+c.status+"] "+c.statusText))},c.setRequestHeader("X-Requested-With","XMLHttpRequest"),c.send(null)):k.test(d)?(d=d.split(":")[1],h=REDIPS.dialog.youtube.replace(/_youtubeId_/g,d),m(h)):m(h+("<div>"+d+"</div>"),q,r);REDIPS.dialog.event.displayed()},hide:function(a,b){!0===REDIPS.dialog.shown&&(REDIPS.dialog.shown=!1,l=a,w=b,t(REDIPS.dialog.opHigh,-10),f.style.display="none",f.innerHTML="",REDIPS.dialog.event.closed())},
html:function(a){g.innerHTML=a},event:{displayed:function(){},closed:function(){}}}}();REDIPS.event||(REDIPS.event=function(){return{add:function(c,g,e){c.addEventListener?c.addEventListener(g,e,!1):c.attachEvent?c.attachEvent("on"+g,e):c["on"+g]=e},remove:function(c,g,e){c.removeEventListener?c.removeEventListener(g,e,!1):c.detachEvent?c.detachEvent("on"+g,e):c["on"+g]=null}}}());

redips-dialog-source.js
/*
Copyright (c)  2008-2020, www.redips.net  All rights reserved.
Code licensed under the BSD License: http://www.redips.net/license/
http://www.redips.net/javascript/dialog-box/
Version 1.8.0
Mar 24, 2020.
*/

/* eslint-env browser */
/* eslint
   semi: ["error", "always"],
   indent: [2, "tab"],
   no-tabs: 0,
   no-multiple-empty-lines: ["error", {"max": 2, "maxEOF": 1}],
   one-var: ["error", "always"] */

/* enable strict mode */
'use strict';

/**
 * @name REDIPS
 * @description create REDIPS namespace (if is not already defined in another REDIPS package)
 */
var REDIPS = REDIPS || {}; // eslint-disable-line no-use-before-define

/**
 * @namespace
 * @description REDIPS.dialog is a simple JavaScript dialog box.
 * @name REDIPS.dialog
 * @author Darko Bunic
 * @see
 * <a href="http://www.redips.net/javascript/dialog-box/">JavaScript dialog box</a>
 * @version 1.8.0
 */
REDIPS.dialog = (function () {
	// function declaration
	var	init,
		show,
		hide,
		imageTag,			// prepare image tag
		position,
		fade,
		inputHtml,			// method prepares input tag HTML
		html,				// method sets custom HTML to display in dialog box
		dialogHtml,			// prepares dialog html
		normalize,			// private method returns normalized spaces from input string
		// properties
		xhr,				// XMLHttpRequest object
		opHigh = 60,		// highest opacity level
		opLow = 0,			// lowest opacity level (should be the same as initial opacity in the CSS)
		fadeSpeed = 10,		// set default speed - 10ms
		shown = false,		// (boolean) readonly public property which shows if dialog is displayed or not
		closeButton = '✕',	// (string) define close button
		customHtml,			// custom HTML to display inside dialog box (customHtml is set in REDIPS.drag.html method)
		// youtube HTML code (this can be overwritten with REDIPS.dialog.youtube - "_youtubeId_" must be present)
		youtube = '<iframe width="640" height="390" src="https://www.youtube.com/embed/_youtubeId_" frameborder="0" allowfullscreen></iframe>',

		// private parameters
		shade,							// shade div (object reference)
		dialogBox,						// dialog box (object reference)
		dialogWidth = 0,				// initialize dialog width
		dialogHeight = 0,				// initialize dialog height
		functionCall,					// name of function to call after clicking on button
		functionParam,					// optional function parameter
		dialogId = 'redips-dialog',		// set dialog id (the same id should be in redips-dialog.css)
		// (object) event handlers
		event = {displayed: function () {},
			closed: function () {}};

	/**
	 * Dialog box initialization.
	 * It should be called at least once before execution of REDIPS.dialog.show() method.
	 * @public
	 * @function
	 * @name REDIPS.dialog#init
	 */
	init = function () {
		// create dialog box
		dialogBox = document.createElement('div');
		dialogBox.setAttribute('id', dialogId);
		// create DIV element to pass custom HTML (displayed in dialog box)
		customHtml = document.createElement('div');
		// create shade div
		shade = document.createElement('div');
		shade.setAttribute('id', 'redips-dialog-shade');
		// append dialog box and shade to the page body
		var body = document.getElementsByTagName('body')[0];
		body.appendChild(shade);
		body.appendChild(dialogBox);
		// add event listener for onscroll & onresize events
		REDIPS.event.add(window, 'resize', position);
		REDIPS.event.add(window, 'scroll', position);
		// initiate XML Http request object
		xhr = new XMLHttpRequest();
	};


	/**
	 * Show dialog box. Method will shade page background and display dialog box.
	 * @param {Integer} width Dialog box width.
	 * @param {Integer} height Dialog box height.
	 * @param {String} text It can contain text, image, page (.php .html .aspx) and 'html' as parameter.
	 * @param {String} [button1] Define first button.
	 * @param {String} [button2] Define second button.
	 * @see <a href="#hide">hide</a>
	 * @public
	 * @function
	 * @name REDIPS.dialog#show
	 */
	show = function (width, height, text, button1, button2) {
		let input1 = '',		// define input1 button
			input2 = '',		// define input2 (optional) button
			imgExtensions,		// needed for images search
			pageExtensions,		// needed for pages search
			youtubeUrl,			// needed for youtube url search
			youtubeHtml = '',	// youtube embeded html
			divImg = '',		// prepared DIV with images
			divText = '',		// text wrapped with DIV
			imgText = '';		// text under image

		// set shown property to true
		REDIPS.dialog.shown = true;
		// set dialog width, height and calculate central position
		dialogWidth = width;
		dialogHeight = height;
		position();
		// if text ends with jpg, jpeg, gif or png, then prepare img tag
		imgExtensions = /(\.jpg|\.jpeg|\.gif|\.png)$/i;
		// if text contains .php, .html, .aspx then page will be fetched via AJAX
		pageExtensions = /(\.php|\.html|\.aspx)/i;
		// if text contains youtube:id
		youtubeUrl = /youtube:/i;
		// prepare optional button1
		if (button1 !== undefined) {
			input1 = inputHtml(button1);
		}
		// prepare optional button2
		if (button2 !== undefined) {
			input2 = inputHtml(button2);
		}
		// if input text is 'html' then it be considered as input parameter and HTML from REDIPS.dialog.html object will be used
		if (text === 'html') {
			dialogHtml(customHtml.innerHTML, input1, input2);
		}
		// if text is image
		else if (imgExtensions.test(text)) {
			// text can precede jpg, jpeg, gif or png image, so search for separator
			imgText = text.split('|');
			// separator doesn't exist, display only image without text
			if (imgText.length === 1) {
				divImg = imageTag(imgText[0]);
			}
			// otherwise, prepare image and text DIV
			else {
				divImg = imageTag(imgText[1]);
				divText = '<div>' + imgText[0] + '</div>';
			}
			dialogHtml(divImg + divText, input1, input2);
		}
		// get content with AJAX from server
		else if (pageExtensions.test(text)) {
			// open asynchronus request
			xhr.open('GET', text, true);
			// set callback handler
			xhr.onreadystatechange = function () {
				// if operation is completed (readyState === 4)
				if (xhr.readyState === XMLHttpRequest.DONE) {
					// if the HTTP status is OK
					if (xhr.status === 200) {
						dialogHtml(xhr.responseText);
					}
					// if request status is not OK then display error in dialog
					else {
						dialogHtml('Error: [' + xhr.status + '] ' + xhr.statusText);
					}
				}
			};
			// in a good manners, set 'X-Requested-With' HTTP header
			xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
			// send request
			xhr.send(null);
		}
		// prepare youtube HTML code (text is in format 'youtube:12345678')
		else if (youtubeUrl.test(text)) {
			// get youtube id
			let id = text.split(':')[1];
			// replace 'youtube' property with id
			youtubeHtml = REDIPS.dialog.youtube.replace(/_youtubeId_/g, id);
			// display HTML
			dialogHtml(youtubeHtml);
		}
		// else prepare text within DIV
		else {
			divText = '<div>' + text + '</div>';
			dialogHtml(divImg + divText, input1, input2);
		}
		// call event handler
		REDIPS.dialog.event.displayed();
	};


	/**
	 * Method sets innerHTML for displaying text, images, youtube videos or custom HTML inside dialog box.
	 * "html" parameter is mandatory while input1 and input2 are optional and they will be mostly used
	 * for displaying text (questions).
	 * @param {String} htmlDialog HTML code to display in dialog box.
	 * @param {String} [input1] First button.
	 * @param {String} [input2] Second button.
	 * @private
	 * @memberOf REDIPS.drag#
	 */
	dialogHtml = function (htmlDialog, input1, input2) {
		// if buttons are not defined, set blank string to enable string concatenation
		if (input1 === undefined) {
			input1 = '';
		}
		if (input2 === undefined) {
			input2 = '';
		}
		// set HTML for dialog box - use table to vertical align content inside
		// dialog box (this should work in all browsers)
		dialogBox.innerHTML =
			'<div class="redips-dialog-titlebar"><span title="Close" onclick="REDIPS.dialog.hide(\'undefined\')">' + REDIPS.dialog.closeButton + '</span></div>' +
			'<table class="redips-dialog-tbl" cellpadding="0" cellspacing="0"><tr><td valign="center" height="' + dialogHeight + '" width="' + dialogWidth + '">' +
			htmlDialog +
			'<div class="redips-dialog-buttons">' + input1 + input2 + '</div>' +
			'</td></tr></table>';
		// show shade and dialog box
		shade.style.display = dialogBox.style.display = 'block';
		// show shaded div slowly
		fade(REDIPS.dialog.opLow, 5);
	};


	/**
	 * Hide dialog box and remove shade (fade out page).
	 * @param {String} fnc Set function call on dialog box hide. If function is not needed then input parameter should be string "undefined".
	 * @param {String} [param] Set parameter to send to the function.
	 * @public
	 * @function
	 * @name REDIPS.dialog#hide
	 */
	hide = function (fnc, param) {
		// test if dialog is displayed
		if (REDIPS.dialog.shown === true) {
			// set shown property to false
			REDIPS.dialog.shown = false;
			// set function call
			functionCall = fnc;
			// set function parameter
			functionParam = param;
			// start fade out
			fade(REDIPS.dialog.opHigh, -10);
			// hide dialog box
			dialogBox.style.display = 'none';
			// clear HTML
			dialogBox.innerHTML = '';
			// call event handler after dialog is hidden (or better say closed)
			REDIPS.dialog.event.closed();
		}
	};


	/**
	 * Method prepares input tag HTML based on "Yes|button2|hello" syntax.
	 * @param {String} button Input string in a form "Yes|button2|hello".
	 * @return {String} Returns HTML of "input" button with prepared REDIPS.dialog.hide() call. This HTML will be used to display buttons in dialog box.
	 * @private
	 * @memberOf REDIPS.dialog#
	 */
	inputHtml = function (button) {
		let param,		// optional parameter for function hide
			inputTag;	// input tag HTML
		// split button values
		button = button.split('|');
		// define parameter (this is last value in composed string)
		param = button[2];
		// prepare optional function parameter
		if (param !== undefined) {
			param = '\',\'' + param;
		}
		else {
			param = '';
		}
		// prepare input tag HTML
		inputTag = '<input type="button" onclick="REDIPS.dialog.hide(\'' + normalize(button[1]) + normalize(param) + '\');" value="' + normalize(button[0]) + '"/>';
		// return result
		return inputTag;
	};


	/**
	 * If text parameter in REDIPS.dialog.show() method contains image (.jpg .jpeg .gif .png) then this method will prepare complete HTML (image wrapped inside DIV element).
	 * @param {String} image Image to display (it is text like "my_image.jpg).
	 * @return {String} Returns HTML for image prepared to display in dialog box.
	 * @private
	 * @memberOf REDIPS.dialog#
	 */
	imageTag = function (image) {
		let img,	// prepared img HTML
			images,	// array containing separated images
			i;		// variable used in loop
		// split input image parameter separated with ,
		images = image.split(',');
		// array contain only one image - simple
		if (images.length === 1) {
			img = '<div class="redips-dialog-imgc"><img src="' + images[0] + '" height="' + (dialogHeight - 40) + '"/></div>';
		}
		// otherwise run loop for more images (images are placed in a table row)
		else {
			img = '<div class="redips-dialog-imgc" style="width:' + (dialogWidth - 8) + 'px"><table><tr>';
			for (i = 0; i < images.length; i++) {
				img += '<td><img src="' + images[i] + '" height="' + (dialogHeight - 40) + '"/></td>';
			}
			img += '</tr></table></div>';
		}
		// return prepared img HTML
		return img;
	};


	/**
	 * Method sets HTML to the customHtml object.
	 * Content of "customHtml" will be used if REDIPS.dialog.show() method will have "html" as string parameter.
	 * @param {String} my_html Custom HTML to show in dialog box.
	 * @see <a href="#show">show</a>
	 * @public
	 * @function
	 * @name REDIPS.dialog#html
	 */
	html = function (myHTML) {
		customHtml.innerHTML = myHTML;
	};


	/**
	 * Method sets dialog position to the center and maximize shade div.
	 * @private
	 * @memberOf REDIPS.dialog#
	 */
	position = function () {
		// define local variables
		let windowWidth,
			windowHeight,
			scrollX,
			scrollY;
		// Non-IE (Netscape compliant)
		if (typeof (window.innerWidth) === 'number') {
			windowWidth = window.innerWidth;
			windowHeight = window.innerHeight;
			scrollX = window.pageXOffset;
			scrollY = window.pageYOffset;
		}
		// IE 6 standards compliant mode
		else if (document.documentElement && (document.documentElement.clientWidth || document.documentElement.clientHeight)) {
			windowWidth = document.documentElement.clientWidth;
			windowHeight = document.documentElement.clientHeight;
			scrollX = document.documentElement.scrollLeft;
			scrollY = document.documentElement.scrollTop;
			// IE hack because #shade{width:100%;height:100%;} will not work for IE if body{height:100%} isn't set also ?!
			shade.style.width = windowWidth + 'px';
			shade.style.height = windowHeight + 'px';
		}
		// DOM compliant
		else if (document.body && (document.body.clientWidth || document.body.clientHeight)) {
			windowWidth = document.body.clientWidth;
			windowHeight = document.body.clientHeight;
			scrollX = document.body.scrollLeft;
			scrollY = document.body.scrollTop;
		}
		// place dialog box to the center (dialog box has "position: fixed" so scrollX & scrollY are not needed for calculation)
		dialogBox.style.left = ((windowWidth - dialogWidth) / 2) + 'px';
		dialogBox.style.top = ((windowHeight - dialogHeight) / 2) + 'px';

		// set shade offset
		shade.style.top = scrollY + 'px';
		shade.style.left = scrollX + 'px';
	};


	/**
	 * Method fades in / fades out page background.
	 * If step is negative then method will fade out otherwise it will fade in (shade) page background.
	 * @param {Integer} opacity Defines opacity level for fade in / fade out.
	 * @param {Integer} step Defines step for fade in / fade out process. Step can be positive or negative.
	 * @private
	 * @memberOf REDIPS.dialog#
	 */
	fade = function (opacity, step) {
		// set opacity for FF and IE
		shade.style.opacity = opacity / 100;
		shade.style.filter = 'alpha(opacity=' + opacity + ')';
		// update opacity level
		opacity += step;
		// recursive call if opacity is between 'low' and 'high' values
		if (REDIPS.dialog.opLow <= opacity && opacity <= REDIPS.dialog.opHigh) {
			setTimeout(
				function () {
					fade(opacity, step);
				}, REDIPS.dialog.fadeSpeed); // fade speed is public parameter
		}
		// hide shade div when fade out ends and make function call
		else if (REDIPS.dialog.opLow > opacity) {
			shade.style.display = 'none';
			if (functionCall !== 'undefined') {
				// call function after button is clicked because functions are defined in global scope
				window[functionCall](functionParam);
			}
		}
	};


	/**
	 * Method returns a string in which all of the preceding and trailing white space has been
	 * removed, and in which all internal sequences of white is replaced with one white space.
	 * @param {String} str Input string.
	 * @return {String} Returns normalized string.
	 * @private
	 * @memberOf REDIPS.dialog#
	 */
	normalize = function (str) {
		if (str !== undefined) {
			str = str.replace(/^\s+|\s+$/g, '').replace(/\s{2,}/g, ' ');
		}
		// return normalized string (without preceding and trailing spaces)
		return str;
	};


	return {
		// public properties
		opHigh: opHigh,			// highest opacity level
		opLow: opLow,			// lowest opacity level (should be the same as initial opacity in the CSS)
		fadeSpeed: fadeSpeed,	// fade speed (default is 18ms)
		youtube: youtube,		// youtube HTML code
		shown: shown,			// (boolean) readonly public property which shows if dialog is displayed or not
		closeButton: closeButton,	// close button (1 character default ✕)

		// public methods
		init: init, // initialization
		show: show, // show dialog box
		hide: hide, // hide dialog box
		html: html, // define custom HTML

		/* Event Handlers */
		/**
		 * All events are part of REDIPS.dialog.event namespace.
		 * @type Object
		 * @ignore
		 */
		event: event
		/* Event Handlers */
		/**
		 * Event handler invoked after dialog is displayed on page.
		 * @param {HTMLElement} [currentCell] Reference to the table cell of clicked element.
		 * @name REDIPS.dialog#event:event::displayed
		 * @function
		 * @event
		 */
		/**
		 * Event handler invoked after dialog is closed.
		 * @name REDIPS.dialog#event:event::closed
		 * @function
		 * @event
		 */
	};
}());


// if REDIPS.event isn't already defined (from other REDIPS file)
if (!REDIPS.event) {
	REDIPS.event = (function () {
		var add,	// add event listener
			remove;	// remove event listener

		// http://msdn.microsoft.com/en-us/scriptjunkie/ff728624
		// http://www.javascriptrules.com/2009/07/22/cross-browser-event-listener-with-design-patterns/

		// add event listener
		add = function (obj, eventName, handler) {
			if (obj.addEventListener) {
				obj.addEventListener(eventName, handler, false);
			}
			else if (obj.attachEvent) {
				obj.attachEvent('on' + eventName, handler);
			}
			else {
				obj['on' + eventName] = handler;
			}
		};

		// remove event listener
		remove = function (obj, eventName, handler) {
			if (obj.removeEventListener) {
				obj.removeEventListener(eventName, handler, false);
			}
			else if (obj.detachEvent) {
				obj.detachEvent('on' + eventName, handler);
			}
			else {
				obj['on' + eventName] = null;
			}
		};

		return {
			add: add,
			remove: remove
		}; // end of public (return statement)
	}());
}
